### ISSUE TITLE
Pod Memory Leak with Gradual Increase

### ISSUE THREAD

**Issue by sinamna** [nats-io/nats-server#6307]
### Observed behavior

One of the cluster pods exhibits a memory leak, with its memory usage gradually increasing over time, even when there is no abnormal activity. In some cases, the memory usage continues to grow despite having no active streams or consumers.

### Expected behavior

The pod's memory usage should remain stable and consistent under normal conditions, without unexpected growth when there are no active streams or consumers.

### Server and client version

server version: 2.10.21

### Host environment

Here are some profiles and logs collected from the pod exhibiting this behavior
[profiles.zip](https://github.com/user-attachments/files/18270291/profiles.zip)
[js-nats-production-2.log](https://github.com/user-attachments/files/18270301/js-nats-production-2.log)
<img width="1502" alt="image" src="https://github.com/user-attachments/assets/0b69c354-6dc4-4e1c-956d-dacb731b8c77" />

**Comment by neilalexander** [comment]
Is this on Kubernetes?

**Comment by lizgbj** [comment]
stream storage = memory ,  memory also increasing  over time ,  server is 2.10.24

**Comment by sinamna** [comment]
> Is this on Kubernetes?

Yes this happens on kubernetes.

**Comment by aialok-famhq** [comment]
@sinamna did you figure out any solution for this issue? or what causing this? We are also facing something similar.
Thanks

**Comment by wallyqs** [comment]
@aialok-famhq this will depend a lot on the type of streams that you have setup, are you using WorkQueue/Interest streams? You can try to narrow down the following things:
- Check with `nats server request raftz` in all accounts whether the WAL size has been gradually increasing in a similar pattern.
- Check with `nats request '$SYS.REQ.SERVER.PING.EXPVARZ'` whether this is just the RSS report and compare with the results from the actual Go runtime.

**Comment by aialok-famhq** [comment]
Hi, thanks for sharing this. I’ll verify it in our system as well.

However, I wanted to clarify a few things.

From what I observe, the NATS server’s memory consumption appears to be stable, so there doesn’t seem to be an issue at the server level.

### NATS Server Memory (Stable)

![NATS Memory](https://github.com/user-attachments/assets/a15ac194-6fc4-4fcd-b5fb-76f40eb12e61)


But the **pods where NATS is running show a different behavior**—memory usage keeps increasing over time, and then around **6:30 AM**, there seems to be some internal cleanup that significantly reduces memory usage.

### Pod Memory Behavior

![Pod Memory](https://github.com/user-attachments/assets/ebc698de-91f5-42df-9661-8db5b2d8a968)


If I look at **only NATS memory consumption**, it stays around **~60%**. However, when I check the **overall pod memory consumption** (which includes NATS, Prometheus, and the config reloader), it goes beyond **120%**.

### NATS Memory Only

![NATS Only](https://github.com/user-attachments/assets/4df95b41-7b7b-491f-929e-5f9aae5c7149)


When isolating other components inside the pod (excluding NATS), they seem to be consuming a significant amount of additional memory.

### Non-NATS Components Memory

![Other Components](https://github.com/user-attachments/assets/f5e3a4f6-5902-4280-bc5e-8aa205a394f0)


So, I’d like to understand what exactly is contributing to this **extra ~60% memory usage**.

Here is our current configuration:

```yaml
- name: GOMEMLIMIT
  value: 4000MiB

resources:
  limits:
    cpu: 2000m
    memory: 4000Mi
  requests:
    cpu: 2000m
    memory: 4000Mi
```


Also, we are using **JetStream with external PVC**. I came across recommendations suggesting that relying on **pod memory instead of disk (PVC)** might be better in some cases.

Additionally, I found some discussions about **page cache behavior**, which seem relevant here. Could you please explain in more detail how page cache works in this context?
https://github.com/nats-io/nats-server/issues/6004#issuecomment-2449623893

Specifically:

* Are reads happening from the **PVC (disk)** or mostly from the **page cache (memory)**?
* Could page cache be responsible for the additional memory usage we are seeing?


Finally, based on the graphs I’ve shared, does everything look normal, or is there something concerning here?

**Comment by wallyqs** [comment]
you need to check the expvarz to be able to tell what is happening

**Comment by aialok-famhq** [comment]
Hi @wallyqs , sharing our profiling results for review:
Profiles: https://github.com/user-attachments/files/26189298/profiles.zip

- We are seeing very high allocations (TB scale) but low in-use memory (~70–90 MB) → suggests allocation churn, not a leak.
- ~80–85% of allocations are from (*consumerFileStore).stateWithCopyLocked.
- Call path observed: checkInterestState → needAck → BorrowState → stateWithCopy.
- Setup uses interest retention + explicit ACK + high throughput (max_ack_pending ~100k).
- Suspect frequent needAck() calls triggering full consumer state copies, causing high GC pressure.

Could you please check the profiling at your end and suggest optimizations? This would help us validate and proceed further.

Thank you so much for your time and consideration : )

**Comment by neilalexander** [comment]
Which NATS Server version?

**Comment by aialok-famhq** [comment]
@neilalexander  this is what we are using - `image: nats:2.11.6-alpine`
Please let me know if u need more information.
Thank you so much for your time and consideration : )

**Comment by aialok-famhq** [comment]
@wallyqs It looks like we are facing a similar issue based on the profiler details: [https://github.com/nats-io/nats-server/issues/6307#issuecomment-4074337781](https://github.com/nats-io/nats-server/issues/6307#issuecomment-4074337781)

Also, on our side, we found a bug where one of our consumer is not acknowledging messages. This is causing messages to be redelivered four times and then dropped.

Really appreciate your help with our profiler.
thank you : )

**Comment by aialok-famhq** [comment]
hi team, really appreciate your help here.
we are seeing gradual increase and memory and suddenly pods restart.
cc : @neilalexander  @wallyqs 
thank you!

**Comment by 1995parham** [comment]
Coming back to this 17 months after @sinamna's original report, with cross-cluster data confirming the pattern is still present and has now traversed a few version lines. Posting in case the breadth is useful — happy to capture `$SYS.REQ.SERVER.PING.EXPVARZ` and `nats server request raftz` on the next occurrence (per @wallyqs's earlier suggestions) if the team wants targeted data.

### Fleet snapshot (taken 2026-05-21)

Two 5-node JetStream clusters (Cluster A, Cluster B) and several 3-node ones, all `nats-server` with file-backed streams, no memory-backed streams in play:

| Cluster | Version | Pods | Pod RSS spread | Subs/pod | Uptime |
| --- | --- | --- | --- | --- | --- |
| A (busy) | 2.14.0 | 5 | 1.0 – **12 GiB** (hit cgroup limit) | ~7,500 | 8d |
| B (less busy) | 2.14.0 | 5 | 0.6 – 1.4 GiB | ~4,200 | 8d |
| C | 2.11.6 | 3 | 27 – 38 MiB | ~250 | 189d |
| D | 2.11.6 | 3 | 124 – 207 MiB | ~660 | 100d |
| E | 2.11.8 | 3 | 156 – 176 MiB | ~520 | 157d |
| F | 2.11.6 | 1 | 194 MiB | 1,660 | 182d |
| G | 2.11.6 | 3 | 22 – 23 MiB | ~330 | 100d |
| H | 2.11.8 | 3 | 508 – 634 MiB | ~3,360 | 2d (recent reroll) |

The same operational pattern @sinamna described originally on 2.10.21, now reproduced on 2.14.0 with much higher slope, and observed historically on the 2.11.x line at a slower rate. Each version line has a different upstream contributor; the symptom is consistent.

### What the affected pod looked like (Cluster A, 2.14.0)

- Process RSS **12 GiB** (= container limit); `Memory In Use` (JetStream's own accounting) **9.6 MiB**; `File In Use` 3.5 GiB on disk.
- Wide spread across the 5-node cluster (12 / 5.4 / 3.3 / 2.5 / 1.0 GiB) with **uniform** subs (~7.5k) and stream count (7–10), so this isn't workload-driven.
- The affected pod was actually *lighter* on consumers (29 vs 145–161 on peers) and API requests (319 vs up to 1.44M) — i.e. fewer code paths running, yet most RSS.
- 0 container restarts in the 8 days of uptime.

### Mitigation behaviour (consistent with the matter being in nats-server process state, not data)

A single-pod `kubectl delete pod js-nats-production-N` brought RSS from 12 GiB → **414 MiB** in under a minute on the recreated pod, with the same Raft ID retained and the meta group `current=true, lag=0` on rejoin. Restart restores the baseline; over the next ~7 days RSS climbs again. We've operationally treated this as a periodic restart for years across version lines.

### Relation to recent upstream fixes

[#8186 "Snapshot and catchup loops no longer leak timers"](https://github.com/nats-io/nats-server/pull/8186) (in 2.14.1 / 2.12.9) appears to be the *2.14.0-specific* contributor — leak rate dropped sharply from anything we'd seen on 2.11.x — but the underlying class of "RSS gradually grows on JS pods over days with no corresponding `Memory In Use` growth" has been present across 2.10.x, 2.11.x, and now 2.14.0. We expect 2.14.1 to make this much slower, not absent.

### Offer

Next time the alert fires on one of our 2.11.x or 2.14.x clusters, I can capture:

- `$SYS.REQ.SERVER.PING.EXPVARZ` (Go runtime memstats vs RSS)
- `nats server request raftz` (per-Raft-group WAL size trend)
- `/varz`, `/jsz`, `/goroutinez`, `/profilez?profile=heap` from the worst-affected pod

Let me know which of these would be most useful to drop here, and whether you'd like me to break it out into a new issue scoped to the post-2.14.1 residual or keep it in this thread.

— happy to provide more context off-issue if useful.

**Comment by neilalexander** [comment]
@1995parham If you can provide memory profiles (https://docs.nats.io/running-a-nats-service/nats_admin/profiling) then that would help. What we have seen often in the past is that Kubernetes pod memory reporting doesn't match the genuine process RSS, mostly down to the fact that Kubernetes counts the page cache into the accounting. If there's evidence that's not what's happening here then we are happy to take a look, but will need profiling data to do so.

**Comment by 1995parham** [comment]
Hi @neilalexander, thanks — happy to provide profiles. I've grabbed a current-state bundle from a `nats-server` 2.14.0 (`go1.26.2`) JetStream cluster and will re-run the same capture next time `NatsMemoryPressure` actually fires; this particular cluster is at ~15% of its container limit right now, but the heap profile already shows what's holding the memory.

Bundle attached (see `SUMMARY.md` inside for the full table). Quick numbers from the two largest nodes of a 5-node R3 JetStream cluster, both 8d uptime, 12 GiB container limit:

| Node | `oc top` RSS | Go `MemStats.Sys` | `HeapInuse` | `HeapAlloc` | `HeapIdle` | `HeapReleased` |
|---|---:|---:|---:|---:|---:|---:|
| -3 | 1865 MiB | 1655 MiB | 917 MiB | 779 MiB | 584 MiB | **38 MiB** |
| -4 | 1228 MiB | 1194 MiB | 822 MiB | 788 MiB | 173 MiB |  **1.4 MiB** |

So this isn't the page-cache accounting case — Go's reported `Sys` tracks the cgroup RSS within ~210 MiB. The Go runtime is holding the heap rather than returning it (`HeapReleased` near zero while `HeapIdle` is large), consistent with sustained churn.

`go tool pprof -top -inuse_space` on the heap profiles of **both** bloated nodes looks identical:

```
       flat  flat%   sum%        cum   cum%
  304.35MB 63.53% 63.53%   348.58MB 72.76%  (*Account).addRespServiceImport
   50.50MB 10.54% 74.07%   399.08MB 83.31%  (*client).setupResponseServiceImport
   44.23MB  9.23% 83.31%    44.23MB  9.23%  (*Account).addReverseRespMapEntry
   27.17MB  5.67% 88.98%    28.67MB  5.98%  (*Server).createRoute
```

~83% of inuse heap is service-import response-map state on the `processInboundClientMsg → setupResponseServiceImport → addRespServiceImport` path; another ~9% in `addReverseRespMapEntry`. Same pattern on `node-4` even though it currently hosts **zero** raft groups in the user account, which to me rules out raft/JS state as the cause on that node.

Raft WAL is also small (≤13 MiB total per pod cluster-wide), so this doesn't look like the WAL-accumulation suspect @wallyqs mentioned earlier in the thread either.

The bundle includes:

- `expvarz_all.txt` — `$SYS.REQ.SERVER.PING.EXPVARZ` for all 5 servers
- `raftz_all.txt` (meta) + `raftz_account_G.txt` (per-account)
- `pod3_varz.json` from HTTP monitor (worst node; host/cluster names redacted)
- `pod3_heap.pb.gz`, `pod3_goroutine.pb.gz`, `pod3_allocs.pb.gz`
- `pod4_heap.pb.gz`, `pod4_goroutine.pb.gz`
- `SUMMARY.md` with the full per-node table

`/debug/pprof` isn't enabled on these nodes (`varz.profile_port = null`), so the profiles are from `$SYS.REQ.SERVER.PING.PROFILEZ`. Let me know if you'd like a longer (`seconds > 0`) CPU profile or anything else, and I'll capture it on the next event.

[nats-6307-capture-2026-05-21.tar.gz](https://github.com/user-attachments/files/28097781/nats-6307-capture-2026-05-21.tar.gz)

**Comment by 1995parham** [comment]
Follow-up — I traced where the inuse heap is coming from. The `addRespServiceImport` allocations all sit on the **system account's** `exports.responses` map, and the source is the **auto-injected** `$JS.API.>` import that NATS creates for every non-system account with JetStream enabled.

`/accountz?acc=%24G` confirms the import (this entry is **not** in our config — we only configure `$SYS` and `system_account`):

```json
{
  "subject": "$JS.API.>",
  "account": "$SYS",
  "to": "$JS.API.>",
  "local_subject": "$JS.API.>",
  "type": "service",
  "share": true
}
```

Created by `server/jetstream.go:737-746`([v2.14.0](https://github.com/nats-io/nats-server/blob/v2.14.0/server/jetstream.go#L737-L746)):

```go
if !a.serviceImportExists(dstAccName, jsAllAPI) {
    si, err := a.addServiceImport(s.SystemAccount(), jsAllAPI, jsAllAPI, nil)
    ...
    si.share = true
}
```

`/accountz?acc=%24SYS` shows ≥488,154 live entries in `$SYS.exports.responses` (response was truncated at 28 MiB by my client, real number is larger). All have `"account": "$G"` and a `_R_.<hash>.<id>` → `_INBOX.<…>` mapping — i.e. every entry is a cross-account JS API request whose reply is still being tracked.

This matches the inline TODO at `server/client.go:4788-4791` ([v2.14.0](https://github.com/nats-io/nats-server/blob/v2.14.0/server/client.go#L4788-L4791)):

```go
// We have a special case where JetStream pulls in all service imports through one export.
// However the GetNext for consumers and DirectGet for streams are a no-op and causes buildups of service imports,
// response service imports and rrMap entries which all will need to simply expire.
// TODO(dlc) - Come up with something better.
```

GetNext / DirectGet already short-circuit `setupResponseServiceImport` (lines 4779-4793), so they're not the source of the buildup here. The remainder is the rest of `$JS.API.*` — `STREAM.INFO`, `CONSUMER.INFO`, `STREAM.MSG.GET`, the various `*.CREATE` / `*.UPDATE` calls, etc. — none of which are skipped.

Correction to my earlier comment: the `"type": 2` field in accountz is `jwt.Service` (the import-kind), not `ServiceRespType.Chunked`. The auto-export is registered via `AddServiceExport(jsAllAPI, nil)` which defaults to **Singleton** (`accounts.go:2052`). Which raises the actual puzzle: with Singleton, `client.go:5001` says

```go
shouldRemove := !didDeliver || (isResponse && (si.rt == Singleton || len(msg) == LEN_CR_LF))
```

— so a delivered Singleton response *should* trigger `removeRespServiceImport` immediately. The fact that there are ~488 K live entries means either:

1. there's a cleanup gap on response delivery for some subset of JS API calls (response service import is created but never removed), and entries only exit via the 2-minute `respThresh` TTL, or
2. the request rate genuinely is so high that even sub-second round-trips leave ~488 K in flight at any moment.

If (2), back-of-envelope math says ~4 K `$JS.API.*` calls/sec sustained with `respThresh = 2m`, which is plausible for a busy JetStream cluster but I would not have expected it to fully populate the map. If anyone on the team can point at which subset of JS API call types is most likely to miss the response-cleanup path, I'm happy to instrument and confirm.

Either way, this is a single-cluster reproduction of the dlc TODO with concrete numbers. Happy to draft a PR for whichever direction makes sense — possibilities I see:

- broaden the existing GetNext/DirectGet skip to all `$JS.API.>` calls when the requestor's account is the auto-imported one (avoids creating the response SI in the first place, since the reply can flow via the regular inbox routing);
- give the JS auto-export a much shorter `respThresh` (current default is 2 m, which is wildly conservative for JS API replies that almost always land in <100 ms);
- skip auto-importing `$JS.API.>` entirely when the configured topology doesn't need it (e.g. single non-system account with JS native — no domain, no leaf, no multi-account routing).

Which direction would you prefer I work toward?

**Comment by 1995parham** [comment]
Update with a clearer reproduction. I took two `/accountz?acc=%24SYS` snapshots from the same node-3 (worst-RSS pod from the earlier bundle):

| time | `$SYS.exports.responses` size |
|---|---:|
| ~10:33 UTC (heap-profile-implied) | ~1.2 M entries |
| 12:40 UTC (T0, full accountz) | 911 510 entries |
| 12:49 UTC (T1, full accountz) | 1 743 045 entries |

The 911 K vs 1.74 M numbers between T0 and T1 are real *map sizes at those instants* — but I want to flag that the apparent **+832 K-in-9-min delta is partly an artifact of how slow the second fetch was**. Each `accountz?acc=%24SYS` response is 250-500 MB of JSON (one entry per row of `$SYS.exports.responses`), so the second `wget` ran for several minutes; during that time the map kept growing under live traffic. Sustained growth — comparing the heap-profile-implied baseline at 10:33 against the T1 count at 12:49 — is closer to **~67 entries/s net**.

Grouping the entries by inbox prefix (= per-connection 12-char NUID) still cleanly identifies one noisy client:

| Inbox NUID prefix | T0 | T1 | rough share |
|---|---:|---:|---:|
| `O0bg0EH1FTk4` | 748 681 | 1 431 749 | **~82 %** |
| `cW5PcZJw2FWK` | 90 195 | 172 139 | ~10 % |
| `1fF3HT7UonH7` | 28 598 | 54 994 | ~3 % |
| `qlw21BxQNI5j` | 22 033 | 42 182 | ~2 % |
| `tbqBZ1oKMLGb` | 22 003 | 41 981 | ~2 % |

That dominant connection is a Go service using the modern `github.com/nats-io/nats.go/jetstream` package (`v1.47.0`) running `consumer.Consume(handler, jetstream.PullExpiry(10s))` with default `apiPrefix = "$JS.API."`. The published subject for the pull loop is `$JS.API.CONSUMER.MSG.NEXT.<stream>.<consumer>`, with the inbox returned by `p.js.conn.NewInbox()` as the reply.

That subject *should* be caught by the existing skip at [`server/client.go:4779-4793`](https://github.com/nats-io/nats-server/blob/v2.14.0/server/client.go#L4779-L4793) via the `jsRequestNextPreB` prefix check. But the evidence says it isn't:

I cross-referenced the noisy client's currently-subscribed inboxes (from `/connz?subs=true`) against the entries in `/accountz?acc=%24SYS`, and at least one currently-active sub — `_INBOX.O0bg0EH1FTk48FU8S2yLkI` — appears verbatim as the `subject` value of a response-service-import entry in `$SYS.exports.responses`.
That entry can only exist if `setupResponseServiceImport` (`server/client.go:4808`) ran for this call, which means `processServiceImport` got past the skip at line 4792 for a pull-NEXT publish. So the skip is silently no-op for at least this code path.

Walking through the conditions on line 4792:

```go
if shouldReturn || (checkJS && si.se != nil && si.se.acc == c.srv.SystemAccount()) {
    return false
}
```

For the skip to *not* fire on a pull NEXT, one of `checkJS`, `si.se != nil`, or `si.se.acc == sysacct` must be false. My best guesses, in order of plausibility:

1. **`si.se` resolution race.** The auto-import is created in `enableAllJetStreamServiceImportsAndMappings` via `a.addServiceImport(s.SystemAccount(), jsAllAPI, jsAllAPI, nil)`, which resolves `se` at creation time (`server/accounts.go:2061-2066`). If anything causes the `$JS.API.>` export on `$SYS` to be re-created or replaced *after* the import was first installed (account-claim update, JS reload, system account swap, JS domain change?), the import's cached `se` pointer would either be stale or `nil`, and the skip silently stops firing for that account.
2. **A second, non-NEXT JS API call sharing the inbox.** The `jetstream.Consume` runtime issues control messages besides the batch `CONSUMER.MSG.NEXT` (status pings, batch-completion, possibly heartbeats). If any of those publish to a `$JS.API.*` subject that *doesn't* match either `jsDirectGetPreB` or `jsRequestNextPreB`, each one creates a response-import entry on the same inbox.
3. **Something simpler I'm missing.** I'm reasoning from v2.14.0 code; if there's a path in `processServiceImport` that reaches `setupResponseServiceImport` *without* going through the skip block, it'd produce exactly this symptom.

Operationally the cluster is in no immediate distress — node-3 RSS recovered to ~1.7 GiB once the `accountz` fetches stopped chewing heap, and the dedicated `NatsMemoryPressure` SOP we have (restart the JS pod) keeps the symptom bounded over weeks. But the underlying skip-not-firing behavior is the part that's reproducible, and it'd be nice to track down. Happy to add instrumentation, grab a CPU profile, or scale that one consumer down to confirm causation — let me know which would help most.

**Comment by 1995parham** [comment]
Follow-up data point on a second cluster — independent reproduction of the heap shape we posted yesterday from the first cluster.

### Setup

- nats-server **2.14.0** (same as the earlier capture), 5-node JetStream cluster on Kubernetes.
- StatefulSet uptime **9d 14h**, 0 restarts. JetStream account `$G`, system account `$SYS`. Single-account config; `$JS.API.>` cross-account service import is the auto-injected one created by `enableAllJetStreamServiceImportsAndMappings`.
- Workload mix: 7 streams (largest 2.8 GiB / 25 M msgs on file storage), 139 consumers on this pod, 7,440 subscriptions, ~13.4 B `in_msgs` over uptime (~16 k msgs/s sustained), routes pool size 1024.
- **The bloated pod holds zero stream leaders** — so the routing tax is purely from `$JS.API.*` traffic landing on it as a peer, not from being the API target.

### Heap (inuse_space)

`$SYS.REQ.SERVER.<id>.PROFILEZ` with `{"name":"heap"}`, decoded and inspected with `go tool pprof -top -inuse_space -unit=mb`:

```
Showing nodes accounting for 5554.29MB, 95.49% of 5816.50MB total
      flat  flat%   sum%        cum   cum%
 3474.19MB 59.73% 59.73%  4376.41MB 75.24%  (*Account).addRespServiceImport
  902.22MB 15.51% 75.24%   902.22MB 15.51%  (*Account).addReverseRespMapEntry
  697.53MB 11.99% 87.23%  5073.94MB 87.23%  (*client).setupResponseServiceImport
  106.01MB  1.82% 89.06%  5267.25MB 90.56%  (*client).processMsgResults
   ...
```

**~5074 MB (87.2 % of inuse heap) on the same three frames as the earlier capture**, in the same relative order. Earlier capture was ~83 % across the same three frames at ~480 MB total inuse; this one is the same shape an order of magnitude up. The cum tree resolves through the expected path `readLoop → parse → processInboundClientMsg → processMsgResults → deliverMsg → processServiceImport → setupResponseServiceImport → addRespServiceImport`.

### Memstats vs container RSS

EXPVARZ across all five servers:

| pod | Go.Sys | container RSS | ratio | NumGC | uptime |
| --- | ---: | ---: | ---: | ---: | --- |
| 0   | 9672 MB | 10462 MB |  93 % | 60382 | 9d |
| 1   | 3672 MB |  4460 MB |  82 % | 88240 | 9d |
| 2   | 6907 MB |  7685 MB |  90 % | 91126 | 9d |
| 3   |  612 MB |   574 MB | ~100 % |  3709 | 31h (post-restart) |
| 4   | 9297 MB |  9305 MB | ~100 % | 18211 | 9d |

`Go.Sys` ≈ container RSS on every bloated pod → bytes are in the Go heap, not cgo/mmap. NumGC on pod-4 is anomalously low for 9d uptime, but that doesn't change the conclusion.

### Other diagnostic numbers (rule-outs)

- Raft WAL across all `$G` groups on the worst pod: **11 MB** total over 144 groups (raftz `--account '$G'`). Not the cause.
- JetStream file store on the worst pod: **3.6 GB** for 25.8 M messages. ~6 GB gap between disk storage and Go heap is what the response-service-import map fills.
- Goroutines: 8955 total; 4,156 readLoop + 4,156 writeLoop + 4,100 createRoute (the route count is `pool_size=1024 × 4 peers`, as configured, not a leak).
- `jetstream.stats.memory = 1.1 MB` and `jetstream.stats.store = 3.4 GB` in varz — JetStream's own accounting is tiny, as expected for this contributor.

### Why this matters for the skip-not-firing hypothesis

Yesterday's capture identified a single dominant client driving ~82 % of the response-map growth via the modern `nats.go/jetstream` pull-consumer API — i.e. `$JS.API.CONSUMER.MSG.NEXT.*` publishes that were supposed to be filtered by the skip at `server/client.go:4779-4793` but weren't. Today's capture is on a **different** cluster, **different** pod (one that holds zero stream leaders, so it's purely receiving routed traffic), and the heap signature is byte-for-byte the same shape and proportion. That's strong corroboration that the underlying mechanism is server-side and not specific to the prior cluster's traffic mix.

I did **not** pull `accountz?acc=$SYS` on this pod for the per-connection breakdown — it's at 85 % RSS and that endpoint transiently doubles `varz.mem` while marshalling, which would likely push it over. After the SOP restart, a fresh capture on the same pod (or any other bloated peer in this cluster) can repeat the inbox-prefix triage safely.

### Capture bundle

Available on request: EXPVARZ for all 5 servers, raftz `$G` for all 5, varz + jsz from the worst pod, and `heap`/`goroutine`/`allocs` profiles from the same pod, all captured immediately before the SOP restart.

**Comment by 1995parham** [comment]
Third data point in two days — back on Cluster A (the 2026-05-21 cluster), 48 h after that restart cycle. Posting because the **role of the bloated pod has flipped** relative to my previous follow-up, which I think is useful for triage.

### What's different from the previous two captures

- 2026-05-21 (Cluster A original bundle): bloated pod held many `$JS.API.*` calls **and** was an API hot-spot.
- 2026-05-22 (second cluster follow-up): bloated pod held **zero stream leaders** — pure routed-traffic peer — same heap shape.
- **Today (Cluster A, fresh event):** bloated pod is the API hot-spot, **1,566,637** `API Req` against **6–71** on three of its four peers. Same heap shape again.

The shape reproduces across (a) leader/API-hot-spot pods, (b) followers receiving routed traffic only, and (c) two different clusters. That rules out "this is local to one workload role."

### Cluster A, 2026-05-23 capture (worst pod = node-1)

nats-server `2.14.0`, 5-node R3 JetStream cluster (file storage). node-1 uptime 10 d, 0 restarts. Container RSS **11,353 MiB** at capture; peers 450–700 MiB.

EXPVARZ across all 5 servers:

| pod | Go.Sys | HeapAlloc | HeapInuse | HeapIdle | HeapReleased | NumGC | API Req |
|---|---:|---:|---:|---:|---:|---:|---:|
| 0 |   696 MiB |   303 MiB |   348 MiB |   284 MiB |   209 MiB |  16,594 |     22,324 |
| **1** | **11,134 MiB** | **5,644 MiB** | **6,707 MiB** | **3,990 MiB** | **5 MiB** | **89,312** | **1,566,637** |
| 2 |   703 MiB |   339 MiB |   362 MiB |   267 MiB |   204 MiB |   9,152 |         71 |
| 3 |   646 MiB |   253 MiB |   290 MiB |   223 MiB |    91 MiB |   7,685 |    312,890 |
| 4 |   605 MiB |   239 MiB |   269 MiB |   154 MiB |    44 MiB |   2,624 |         11 |

Container RSS / Go `Sys` agree within 2 % on node-1, so the bytes are still in the Go heap (not page cache).

The NumGC on node-1 (**89 312** vs 2.6 k–16.6 k on peers) is the new detail I want to highlight: GC is running 5–30 × more often, but `HeapReleased = 5 MiB` against `HeapIdle = 3,990 MiB` and `HeapAlloc = 5,644 MiB` — i.e. GC keeps re-scanning the same live entries, can't drop them, and the runtime won't release the idle pages back to the OS. That's consistent with the entries being genuinely held by a long-lived structure (the `$SYS.exports.responses` map from the earlier follow-up), not transient churn.

### Heap (inuse_space)

```
File: nats-server
Build ID: 32cc3f47bc300fef80b699865c8790308325edf0
Type: inuse_space
Showing nodes accounting for 5202.65MB, 94.10% of 5528.94MB total
      flat  flat%   sum%        cum   cum%
 3530.89MB 63.86% 63.86%  4135.98MB 74.81%  (*Account).addRespServiceImport
  605.09MB 10.94% 74.81%   605.09MB 10.94%  (*Account).addReverseRespMapEntry
  525.52MB  9.50% 84.31%  4661.50MB 84.31%  (*client).setupResponseServiceImport
  146.14MB  2.64% 86.95%   166.34MB  3.01%  (*raft).sendAppendEntryLocked
  104.01MB  1.88% 88.84%  4830.70MB 87.37%  (*client).processMsgResults
```

Cum tree through the expected path: `readLoop → parse → processInboundClientMsg → processMsgResults → deliverMsg → processInboundMsg → processServiceImport → (*Account).addServiceImportSub.func1 → setupResponseServiceImport → addRespServiceImport / addReverseRespMapEntry`. Same as before.

### Rule-outs (same as previous captures, confirming)

- Raft WAL across all `$G` groups on node-1: **14 MiB** total over 151 groups (`server request raft --account '$G'`). Not the source.
- `jetstream.stats.memory = 8.7 MiB` (file store 5.07 GiB). JetStream's own accounting is tiny.
- Goroutines 9,442 total (4,156 readLoop + 4,156 writeLoop + ~4,100 createRoute — same `pool_size=1024 × peers` shape as before, not a leak).

### Operational note

`kubectl delete pod node-1`: 19 s to Ready, meta-cluster `current=true, lag=0` immediately on rejoin, RSS settled at **529 MiB** within ~30 s. Same SOP that has been working across versions for years.

### Bundle (available on request)

EXPVARZ all 5 servers, raftz `$G` all 5, varz/jsz on node-1, heap/allocs/goroutine profiles on node-1. Captured immediately before the SOP restart.

### What I'd like to do next

Given the GC-can't-release evidence above and the existing inbox-prefix breakdown from the 2026-05-21 capture, the most useful next thing on my side is to scale down the noisy `jetstream.Consume` client identified in that capture and watch whether the map growth slope drops in proportion. If it does, that confirms the skip at `server/client.go:4779-4793` is the right place to fix. Will report back.

**Comment by 1995parham** [comment]
Picking this back up — I think I have the mechanism, and I've opened #8463 with a fix.

Following on from the May captures in this thread: the entries really are on the system account's `exports.responses`, fed by the auto-injected `$JS.API.>` import. The part we could not explain then was why they were not expiring at the 2 minute response threshold. They are. The reaper just cannot keep up.

`Account.imports.rrMap` is keyed by **reply subject** and holds a slice, and removing one entry scans that slice for a matching response subject:

```go
for i, sre := range sres {
    if sre.msub == si.from {
        sres = append(sres[:i], sres[i+1:]...)
        break
    }
}
```

A client is allowed to reuse one reply subject across many concurrent requests, and `nats.go`'s legacy pull subscription `Fetch` does exactly that — `newFetchInbox` returns the subscription's inbox unchanged for a non-wildcard subject. When that happens the slice length tracks the number of outstanding responses, so each removal is O(n) and clearing them all is O(n²).

`checkExpiredResponses` walks every outstanding response per pass and re-arms its timer only once the pass finishes. Past a certain backlog a pass no longer completes inside the threshold, so it falls permanently behind: backlog grows, next pass is slower, memory grows without bound. That matches what we measured — 1.5M entries would need ~12,500 req/s to be steady state at a 2 minute TTL, and we are nowhere near that, so it is a backlog rather than in-flight work.

It also explains the observation from 2026-05-23 that puzzled me at the time, where the bloated pod was sometimes the API hot-spot and sometimes held zero stream leaders. What matters is not how much JetStream work the node does, it is whether a client connected to it reuses a reply subject. On our cluster the nodes with no such client sit at <1 GiB for weeks while the affected one climbs to 10 GiB.

Reproducer numbers, cleaning up n responses sharing one reply subject:

| n | before | after (#8463) |
|---|---|---|
| 5000 | 19ms | 1ms |
| 10000 | 65ms | 1ms |
| 20000 | 219ms | 3ms |
| 40000 | 895ms | 5ms |

Per-entry cost doubles as n doubles today; flat at ~0.1µs with the patch. Extrapolating the old curve to the 1.5M entries we saw in production puts a single cleanup pass in the tens of minutes, against a 2 minute timer.

Worth noting for anyone else hitting this: `$JS.API.CONSUMER.MSG.NEXT` is already exempted in `processServiceImport`, so raw pull volume is not the trigger. In our case it was a client calling `$JS.API.CONSUMER.INFO` on a short interval over a long-lived pull subscription's inbox. Reducing that call rate, or using a client that gives each request its own inbox, avoids the pathological case without a server change.
